### PR TITLE
Go ut

### PR DESCRIPTION
**Pull request by ctwarog-ross**
Added UT framework
Closes Issue #1231

## 📖 Description
Added framework for unit testing Go SDK
90% coverage

---

## ✅ Definition of Done (DoD)
- [x] Feature is implemented and works as expected
- [x] Edge cases are handled
- [x] Tests are added or updated (if applicable)
- [ ] Documentation is updated (if needed)
- [x] Code is reviewed and approved

---

## 🛠️ Implementation Notes
edit to scripts/run_coverage.sh to run tests and coverage
added coverage files to .gitignore
new go library install added to devcontainer

---

## 🔗 Related Issues / Links
https://github.com/rossvideo/Catena/issues/1124

---
## 📝 Additional Context

---

## 🚧 Risks / Open Questions

---
## ✨ Updates
<!-- (Optional) List summarizing changes made when editing the pull request>

### CODE REVIEW

**Review by abrown-RV — CHANGES_REQUESTED** [review]
I'd like the coverage gutters extension to work. It doesn't support the go coverage format directly: https://github.com/ryanluker/vscode-coverage-gutters/issues/303 but the workaround is to add an extra step to coverage to something like lcov or cobertura format which gutters does understand. So just another little bit in run_coverage

Would also like to see the test summary/coverage comment like we have on cpp and js, copy off those/ask Nathan or I and we'd be happy to help.

**Comment by ctwarog-ross** [comment]
> I'd like the coverage gutters extension to work. It doesn't support the go coverage format directly: [ryanluker/vscode-coverage-gutters#303](https://github.com/ryanluker/vscode-coverage-gutters/issues/303) but the workaround is to add an extra step to coverage to something like lcov or cobertura format which gutters does understand. So just another little bit in run_coverage
> 
> Would also like to see the test summary/coverage comment like we have on cpp and js, copy off those/ask Nathan or I and we'd be happy to help.

Created new sub issues under issue#1124 to address your comment Andrew. I'd like to keep every PR singular purposed and this one's scope is already quite large.

File: .devcontainer/go/devcontainer.dockerfile [hunk 1]
```diff
@@ -51,6 +51,8 @@ RUN go install golang.org/x/tools/gopls@latest \
     && go install github.com/go-delve/delve/cmd/dlv@latest \
     && go install google.golang.org/protobuf/cmd/protoc-gen-go@latest \
     && go install google.golang.org/grpc/cmd/protoc-gen-go-grpc@latest
-
+    && go install github.com/axw/gocov/gocov@latest
```
Comment (abrown-RV) [inline .devcontainer/go/devcontainer.dockerfile]:
```suggestion
    && go install google.golang.org/grpc/cmd/protoc-gen-go-grpc@latest \
    && go install github.com/axw/gocov/gocov@latest \
```
dockerfile won't build

feel like these belong in the toolchain, then you don't have to also install them in the workflow

should be version pinned like everything else

File: sdks/go/pkg/logger/settings_test.go
Comment (abrown-RV) [inline sdks/go/pkg/logger/settings_test.go:1]: should this be `settings_test.go`?

File: .github/workflows/go-pull-request.yml [hunk 2]
```diff
@@ -250,3 +116,87 @@ jobs:
...
+      - name: Upload coverage artifacts
+        uses: actions/upload-artifact@v4
+        with:
+          name: go-coverage-report
+          path: |
+            sdks/go/coverage.out
+            sdks/go/coverage.txt
+            sdks/go/coverage.html
+            sdks/go/coverage-detailed.html
+          retention-days: 30
```
Comment (abrown-RV) [inline .github/workflows/go-pull-request.yml]:
```suggestion
          retention-days: 30

```

File: .gitignore [hunk 3]
```diff
@@ -96,3 +96,4 @@ scripts/reports/
 #Go
 *__debug_bin*
 go.work*
+*coverage.out
```
Comment (abrown-RV) [inline .gitignore]:
```suggestion
*coverage.out

```
or just not needed since `coverage/` is already ignored

**Review by abrown-RV — CHANGES_REQUESTED** [review]
Unit tests are hard to review.... the coverage is good so they must be most of the way there. I think an issue to go back and hit some more err branches would be good to get individual files all up higher.

**Review by abrown-RV — DISMISSED** [review]
very cool!

File: sdks/go/internal/testutil/log_capture.go [hunk 4]
```diff
@@ -0,0 +1,180 @@
+/*
+ * Copyright 2025 Ross Video Ltd
```
Comment (NR-RV) [inline sdks/go/internal/testutil/log_capture.go]: its 2026

File: sdks/go/pkg/logger/level_test.go [hunk 5]
```diff
@@ -0,0 +1,127 @@
+/*
+ * Copyright 2025 Ross Video Ltd
```
Comment (NR-RV) [inline sdks/go/pkg/logger/level_test.go]: its 2026

File: sdks/go/pkg/logger/settings_test.go [hunk 6]
```diff
@@ -0,0 +1,276 @@
+/*
+ * Copyright 2025 Ross Video Ltd
```
Comment (NR-RV) [inline sdks/go/pkg/logger/settings_test.go]: its 2026

File: sdks/go/pkg/rest/server_test.go [hunk 7]
```diff
@@ -0,0 +1,931 @@
...
+	rec := httptest.NewRecorder()
+
+	srv.mux.ServeHTTP(rec, req)
+
+	if rec.Code != http.StatusOK {
+		t.Errorf("expected status %d, got %d", http.StatusOK, rec.Code)
+	}
+}
+
+func TestServer_SetValue_InvalidContentType(t *testing.T) {
```
Comment (NR-RV) [inline sdks/go/pkg/rest/server_test.go:362]: on a "should return an error status" test make sure the set is never called
also i dont see the check if the repose is the correct status return

File: sdks/go/pkg/rest/server_test.go [hunk 8]
```diff
@@ -0,0 +1,931 @@
...
+	rec := httptest.NewRecorder()
+
+	srv.mux.ServeHTTP(rec, req)
+
+	if rec.Code != http.StatusOK {
+		t.Errorf("expected status %d, got %d", http.StatusOK, rec.Code)
+	}
+}
+
+func TestServer_ExecuteCommand_MethodNotAllowed(t *testing.T) {
```
Comment (NR-RV) [inline sdks/go/pkg/rest/server_test.go:490]: again on "should return an error status" test make sure the send command is not called

File: sdks/go/pkg/rest/server_test.go [hunk 9]
```diff
@@ -0,0 +1,931 @@
...
+	rec := httptest.NewRecorder()
+
+	srv.mux.ServeHTTP(rec, req)
+
+	if rec.Code != http.StatusOK {
+		t.Errorf("expected status %d, got %d", http.StatusOK, rec.Code)
+	}
+}
+
+func TestServer_Connect_MethodNotAllowed(t *testing.T) {
```
Comment (NR-RV) [inline sdks/go/pkg/rest/server_test.go:535]: see other "should return an error status" tests

File: sdks/go/pkg/rest/server_test.go [hunk 10]
```diff
@@ -0,0 +1,931 @@
...
+	rec := httptest.NewRecorder()
+
+	srv.mux.ServeHTTP(rec, req)
+
+	if rec.Code != http.StatusOK {
+		t.Errorf("expected status %d, got %d", http.StatusOK, rec.Code)
+	}
+}
+
+func TestServer_UnknownEndpoint(t *testing.T) {
```
Comment (NR-RV) [inline sdks/go/pkg/rest/server_test.go:676]: needs to check if the correct status is returned

File: sdks/go/pkg/rest/server_test.go [hunk 11]
```diff
@@ -0,0 +1,931 @@
...
+	req := httptest.NewRequest(http.MethodGet, "/st2138-api/v1/0/unknown", nil)
+	rec := httptest.NewRecorder()
+
+	srv.mux.ServeHTTP(rec, req)
+
+	// Unknown endpoint should return NOT_FOUND in response
+	// Test passes if no panic
+}
+
+func TestServer_InvalidPath(t *testing.T) {
```
Comment (NR-RV) [inline sdks/go/pkg/rest/server_test.go:689]: same as R569

File: sdks/go/pkg/rest/server_test.go [hunk 12]
```diff
@@ -0,0 +1,931 @@
...
+	// Path too short - may redirect or return error
+	req := httptest.NewRequest(http.MethodGet, "/st2138-api/v1", nil)
+	rec := httptest.NewRecorder()
+
+	srv.mux.ServeHTTP(rec, req)
+
+	// Test passes if no panic - path may be redirected by mux
+}
+
+func TestServer_NegativeSlot(t *testing.T) {
```
Comment (NR-RV) [inline sdks/go/pkg/rest/server_test.go:716]: same as R569

File: sdks/go/pkg/rest/server_test.go [hunk 13]
```diff
@@ -0,0 +1,931 @@
...
+
+			body, _ := io.ReadAll(rec.Body)
+			if !bytes.Contains(body, []byte(tt.expected)) {
+				t.Errorf("expected %q in response, got %q", tt.expected, string(body))
+			}
+		})
+	}
+}
+
+func TestWriteFunctions_NilValues(t *testing.T) {
```
Comment (NR-RV) [inline sdks/go/pkg/rest/server_test.go:823]: should also check to make sure no values get changed

File: sdks/go/pkg/rest/server_test.go [hunk 14]
```diff
@@ -0,0 +1,931 @@
...
+	// Test valid asset
+	asset, _ := catena.ToCatenaAsset(catena.DataPayload{Payload: []byte("data")}, false)
+	rec = httptest.NewRecorder()
+	writeAssetResult(rec, asset, http.StatusOK)
+	if rec.Code != http.StatusOK {
+		t.Error("asset write failed")
+	}
+}
+
+func TestCommandEndpoint_InvalidJSON(t *testing.T) {
```
Comment (NR-RV) [inline sdks/go/pkg/rest/server_test.go:1008]: make sure the correct status response is sent

**Review by NR-RV — CHANGES_REQUESTED** [review]
see server tests and check that those comments dont apply to the other tests

File: sdks/go/internal/testutil/log_capture.go [hunk 15]
```diff
@@ -0,0 +1,180 @@
...
+ * and/or other materials provided with the distribution.
+ *
+ * 3. Neither the name of the copyright holder nor the names of its
+ * contributors may be used to endorse or promote products derived from this
+ * software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * RE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
```
Comment (nelgd) [inline sdks/go/internal/testutil/log_capture.go]: says "RE" instead of "ARE"

File: sdks/go/pkg/logger/level_test.go [hunk 16]
```diff
@@ -0,0 +1,127 @@
...
+ * and/or other materials provided with the distribution.
+ *
+ * 3. Neither the name of the copyright holder nor the names of its
+ * contributors may be used to endorse or promote products derived from this
+ * software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * RE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
```
Comment (nelgd) [inline sdks/go/pkg/logger/level_test.go]: Typo here

File: sdks/go/pkg/logger/settings_test.go [hunk 17]
```diff
@@ -0,0 +1,276 @@
...
+ * and/or other materials provided with the distribution.
+ *
+ * 3. Neither the name of the copyright holder nor the names of its
+ * contributors may be used to endorse or promote products derived from this
+ * software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * RE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
```
Comment (nelgd) [inline sdks/go/pkg/logger/settings_test.go]: Typo Here

**Review by nelgd — DISMISSED** [review]

File: sdks/go/pkg/logger/settings_test.go [hunk 18]
```diff
@@ -0,0 +1,276 @@
...
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
+
+/**
+ * @brief Test for Settings and configuration parsing.
+ * @file settings_test.go
+ * @copyright Copyright © 2025 Ross Video Ltd
```
Comment (NR-RV) [inline sdks/go/pkg/logger/settings_test.go]: its 2026

File: sdks/go/pkg/logger/settings_test.go [hunk 19]
```diff
@@ -0,0 +1,276 @@
...
+ * @author Christian Twarog (christian.twarog@rossvideo.com)
+ * @date 2025-12-09
```
Comment (NR-RV) [inline sdks/go/pkg/logger/settings_test.go]: same as r34

**Review by NR-RV — COMMENTED** [review]
small date fix

**Review by NR-RV — APPROVED** [review]
